sunscreen_math/src/opencl_impl/multiexp.rs

```diff
@@ -368,4 +403,65 @@ mod tests {
             assert_eq!(bin_ids[start..end], expected[i].bin_ids);
         }
     }
+
+    #[test]
+    fn can_init_buckets() {
+        for window_size_bits in 1..17 {
+            let num_buckets = multiexp_num_buckets(window_size_bits);
+            let num_windows = multiexp_num_windows(window_size_bits);
+
+            let buckets = init_bucket_points(window_size_bits);
+
+            let buckets = buckets.iter().collect::<Vec<_>>();
+
+            let identity = RistrettoPoint::identity();
+
+            for w in 0..num_windows {
+                for b in 0..num_buckets {
+                    let bucket_point = buckets[b + w * num_buckets];
+
+                    // RistrettoPoint's eq function returns true for invalid points.
+                    // So use something spicier.
+                    assert!(ristretto_bitwise_eq(identity, bucket_point));
+                }
+            }
+        }
+    }
+
+    #[test]
+    fn can_populate_bins() {
+        let count = 4567u32;
+
+        let scalars = (0..count)
+            .map(|_x| Scalar::random(&mut thread_rng()))
+            .collect::<Vec<_>>();
+        let points = (0..count)
+            .map(|_x| RistrettoPoint::random(&mut thread_rng()))
+            .collect::<Vec<_>>();
+
+        let scalars_gpu = ScalarVec::new(&scalars);
+        let points_gpu = RistrettoPointVec::new(&points);
+
+        let window_size_bits = 16;
+
+        let bucket_data = compute_bucket_data(&scalars_gpu, window_size_bits);
+
+        let actual_gpu = compute_bucket_points(&points_gpu, &bucket_data, window_size_bits);
+
+        let actual = actual_gpu.iter().collect::<Vec<_>>();
+
+        let expected = crate::test_impl::compute_bucket_points(&scalars, &points, window_size_bits);
+
+        let num_buckets = multiexp_num_buckets(window_size_bits);
+
+        for (actual_window_buckets, expected_window_buckets) in
+            actual.chunks(num_buckets).zip(expected)
```

Comment:
This does skip testing that `actual.len() / num_buckets == expected.len()`, unsure if that's important.

Reply:
Better safe than sorry.